### ISSUE TITLE
Memory issue with version v3.667.0 

### ISSUE THREAD

**Issue by JeremieDoctrine** [aws/aws-sdk-js-v3#6553]
### Checkboxes for prior research

- [X] I've gone through [Developer Guide](https://docs.aws.amazon.com/sdk-for-javascript/v3/developer-guide) and [API reference](https://docs.aws.amazon.com/AWSJavaScriptSDK/v3/latest)
- [X] I've checked [AWS Forums](https://forums.aws.amazon.com) and [StackOverflow](https://stackoverflow.com/questions/tagged/aws-sdk-js).
- [X] I've searched for [previous similar issues](https://github.com/aws/aws-sdk-js-v3/issues) and didn't find any solution.

### Describe the bug

We have three services using the aws-sdk` v3.667.0
This is the memory usage per version :
![CleanShot 2024-10-09 at 16 01 50@2x](https://github.com/user-attachments/assets/d4190c3d-9689-4390-8752-fa62f8444c3a)

### Regression Issue

- [X] Select this option if this issue appears to be a regression.

### SDK version number

@aws-sdk/client-dynamodb@3.667.0, @aws-sdk/lib-dynamodb@3.667.0

### Which JavaScript Runtime is this issue in?

Node.js

### Details of the browser/Node.js/ReactNative version

v22.4.1

### Reproduction Steps

run `npm start` the container immediately runs out of memory.

### Observed Behavior

Memory increases at boot time

### Expected Behavior

Memory do not increase at boot time

**Comment by JeremieDoctrine** [comment]
I know there is very little information in this ticket. But i'm, quite sure this version is bugged. And wanted to create an issue because i'm sure other people will run into the same issue.

**Comment by Chrisp1tv** [comment]
Hello!
I'm working on a Nodejs (version `22.9.0`) project and I'm observing the same issue when using AWS SDK with NodeJS streams since this version, `3.667.0`. We can provide more details if it can help to debug the issue :)

**Comment by lucavb** [comment]
In our case it is also making our Nest app go OOM on startup.

**Comment by MathieuGuillet** [comment]
Same here with node JS / Nest and these dependencies:
```
    "@aws-sdk/client-s3": "^3.667.0",
    "@aws-sdk/credential-providers": "^3.667.0",
    "@aws-sdk/lib-storage": "^3.667.0",
    "@aws-sdk/lib-storage": "^3.667.0",
    "@aws-sdk/rds-signer": "^3.667.0",
```

**Comment by krukid** [comment]
same here, the memory leak occurred with these aws deps

```
aws-sdk@2.1691.0
@aws-sdk/credential-providers@3.667.0
```

rolling back to `@aws-sdk/credential-providers@3.664.0` fixed the issue

**Comment by kuhe** [comment]
Hi, we are marking 3.666.0 series of `@aws-sdk/*` packages as `latest` on NPM while we prepare a fix.

**Comment by kuhe** [comment]
I'm sorry for not catching this bug. A fix has been made in PR #6555. 

We will release the new version later today.

The root cause is calling the memoized credentials provider function in the user agent middleware. The credentials provider function may include an SDK operation to e.g. STS, and during that invocation it loops into the same middleware.

After resolving the immediate issue, we will investigate how to improve test coverage to avoid recurrences.

**Comment by kuhe** [comment]
https://github.com/aws/aws-sdk-js-v3/releases/tag/v3.668.0 has been released with what I believe is the fix.

That said, does anyone have a more specific reproduction setup for this issue?

**Comment by matt-halliday** [comment]
> https://github.com/aws/aws-sdk-js-v3/releases/tag/v3.668.0 has been released with what I believe is the fix.
> 
> That said, does anyone have a more specific reproduction setup for this issue?

I don't have a full repro repo... can probably create one if needed?

We were seeing it in a node API running in K8s with a service role granting `GetObject` access to a bucket.

A request would come in:
```
/get-foo/{id}`
```

Our `S3Client` would go something like this:
```
try {
    const get = new GetObjectCommand({
        Bucket: 'bucket',
        key: <the-id>,
    });
    // we get here
    const data = await s3Client.send(command);
    return JSON.parse(data.Body);
} catch (err) {
    // we never get here
}
``` 

It would crash immediately...

**Comment by angusjellis** [comment]
We've observed the same behaviour. We have a simple tool that includes the AWS SDK as part of its dependencies. 

It instantiates an S3 client before doing anything else, using an OIDC token from GitLab CI.

When the tool runs on a container with 2GB of RAM, if using version`3.666` it works as expected, but on version `3.667` we can pinpoint that it runs out of memory as soon as the S3 client is instantiated.

**Comment by aqeelat** [comment]
@kuhe `@aws-sdk/types` is not updated. Latest is still 3.664.0.

**Comment by kuhe** [comment]
Not all packages get updated to every version. The clients on version 3.668.0 still use `@aws-sdk/types@3.667.0`, which is not an affected package. The affected package is `@aws-sdk/middleware-user-agent@3.667.0`.

**Comment by kuhe** [comment]
It looks like container credentials may be the precondition, which I'll investigate.

**Comment by matt-halliday** [comment]
> It looks like container credentials may be the precondition, which I'll investigate.

Using a profile, I couldn't recreate locally with a basic benchmark so that's quite possible... I've attached some nasty code if it helps. You can change the client version in `package.json`, run `npm i && npm run benchmark` for output in the console.

[benchmark.zip](https://github.com/user-attachments/files/17330330/benchmark.zip)

**Comment by kuhe** [comment]
I believe this is fixed with v3.668.0, so I'll be closing the issue soon unless anyone reports the issue as persisting in v3.668.0, but I'll comment with the root cause details when I'm able to determine the preconditions.

**Comment by kuhe** [comment]
My testing shows the issue affected `assumeRoleWithWebCredentials` but not `assumeRole`, which usually coincides with usage of `process.env.AWS_WEB_IDENTITY_TOKEN_FILE`.
